segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/filesystem/FileSystemStorage.java

```diff
@@ -0,0 +1,377 @@
+    public FileSystemStorage(FileSystemStorageConfig config, ExecutorService executor) {
+        Preconditions.checkNotNull(config, "config");
+        Preconditions.checkNotNull(executor, "executor");
+        this.config = config;
+        this.executor = executor;
+    }
+
+    //endregion
+
+    //region Storage implementation
+
+    /**
+     * Initialize is a no op here as we do not need a locking mechanism in case of file system write.
+     * @param containerEpoch The Container Epoch to initialize with (ignored here).
+     */
+    @Override
+    public void initialize(long containerEpoch) {
+
+    }
+
+    @Override
+    public CompletableFuture<SegmentHandle> openRead(String streamSegmentName) {
+        return supplyAsync( streamSegmentName, () -> syncOpenRead(streamSegmentName));
+    }
+
+
+    @Override
+    public CompletableFuture<Integer> read(SegmentHandle handle, long offset, byte[] buffer, int bufferOffset, int
+            length, Duration timeout) {
+        return supplyAsync( handle.getSegmentName(), () -> syncRead(handle, offset, buffer, bufferOffset, length));
+    }
+
+    @Override
+    public CompletableFuture<SegmentProperties> getStreamSegmentInfo(String streamSegmentName, Duration timeout) {
+        return supplyAsync(streamSegmentName, () -> syncGetStreamSegmentInfo(streamSegmentName));
+    }
+
+    @Override
+    public CompletableFuture<Boolean> exists(String streamSegmentName, Duration timeout) {
+        return supplyAsync(streamSegmentName, () -> syncExists(streamSegmentName));
+    }
+
+    @Override
+    public CompletableFuture<SegmentHandle> openWrite(String streamSegmentName) {
+        return supplyAsync(streamSegmentName, () -> syncOpenWrite(streamSegmentName));
+    }
+
+    @Override
+    public CompletableFuture<SegmentProperties> create(String streamSegmentName, Duration timeout) {
+        return supplyAsync( streamSegmentName, () -> syncCreate(streamSegmentName));
+    }
+
+    @Override
+    public CompletableFuture<Void> write(SegmentHandle handle, long offset, InputStream data, int length, Duration
+            timeout) {
+        return supplyAsync(handle.getSegmentName(), () -> syncWrite(handle, offset, data, length));
+    }
+
+    @Override
+    public CompletableFuture<Void> seal(SegmentHandle handle, Duration timeout) {
+        return supplyAsync(handle.getSegmentName(), () -> syncSeal(handle));
+    }
+
+    @Override
+    public CompletableFuture<Void> concat(SegmentHandle targetHandle, long offset, String sourceSegment, Duration
+            timeout) {
+            return supplyAsync(targetHandle.getSegmentName(),
+                    () -> syncConcat(targetHandle, offset, sourceSegment));
+    }
+
+    @Override
+    public CompletableFuture<Void> delete(SegmentHandle handle, Duration timeout) {
+            return supplyAsync(handle.getSegmentName(), () -> syncDelete(handle));
+    }
+
+    //endregion
+
+    //region AutoClosable
+
+    @Override
+    public void close() {
+
```

Comment:
Although not strictly necessary, I find it much better to do what @andreipaduroiu is suggesting as it makes the behavior explicit. If it is implementing closable (via transitive relation), then we should implement the behavior properly to comply with the requirements. I'm +1 for @andreipaduroiu 's suggestion and I don't think it would hurt any other behavior in this class, right?